lib/tbot/service_bot_identity.go

```diff
@@ -116,77 +123,99 @@ func (s *identityService) loadIdentityFromStore(ctx context.Context, store bot.D
 			sha := sha256.Sum256([]byte(token))
 			configTokenHashBytes := []byte(hex.EncodeToString(sha[:]))
 			if hasTokenChanged(loadedIdent.TokenHashBytes, configTokenHashBytes) {
-				s.log.InfoContext(ctx, "Bot identity loaded from store does not match configured token. Bot will fetch identity using configured token.")
+				s.log.InfoContext(ctx, "Bot identity loaded from store does not match configured token")
 				// If the token has changed, do not return the loaded
 				// identity.
-				return nil, nil
+				return nil
 			}
 		} else {
 			// we failed to get the newly configured token to compare to,
 			// we'll assume the last good credentials written to disk should
 			// still be used.
-			s.log.ErrorContext(ctx, "There was an error loading the configured token. Bot identity loaded from store will be tried.", "error", err)
+			s.log.WarnContext(
+				ctx,
+				"There was an error loading the configured token to compare to existing identity. Identity loaded from store will be tried",
+				"error", err,
+			)
 		}
 	}
-	s.log.InfoContext(ctx, "Loaded existing bot identity from store.", "identity", describeTLSIdentity(ctx, s.log, loadedIdent))
 
-	return loadedIdent, nil
+	s.log.InfoContext(
+		ctx,
+		"Loaded existing bot identity from store",
+		"identity", describeTLSIdentity(ctx, s.log, loadedIdent),
+	)
+
+	now := time.Now().UTC()
+	if now.After(loadedIdent.X509Cert.NotAfter) {
+		s.log.WarnContext(
+			ctx,
+			"Identity loaded from store is expired",
+			"not_after", loadedIdent.X509Cert.NotAfter.Format(time.RFC3339),
+			"current_time", now.Format(time.RFC3339),
+		)
+		return nil
+	} else if now.Before(loadedIdent.X509Cert.NotBefore) {
+		s.log.WarnContext(
+			ctx,
+			"Identity is not yet valid. Confirm that the system time is correct",
+			"not_before", loadedIdent.X509Cert.NotBefore.Format(time.RFC3339),
+			"current_time", now.Format(time.RFC3339),
+		)
+		return nil
+	}
+
+	return loadedIdent
 }
 
-// Initialize attempts to load an existing identity from the bot's storage.
-// If an identity is found, it is checked against the configured onboarding
-// settings. It is then renewed and persisted.
+// Initialize sets up the bot identity at startup. This process has a few
+// steps to it.
+//
+// First, we attempt to load an existing identity from the configured storage.
+// This is ignored if we know that the onboarding settings have changed.
+//
+// If the identity is found, and seems valid, we attempt to renew using this
+// identity to give us a fresh set of certificates.
 //
-// If no identity is found, or the identity is no longer valid, a new identity
-// is requested using the configured onboarding settings.
+// If there is no identity, or the identity is invalid, we'll join using the
+// configured onboarding settings.
 func (s *identityService) Initialize(ctx context.Context) error {
```

Comment:
One of the key behavioural changes here is that previously, we wouldn't even attempt to load from persisted storage if a bot was configured with a renewable join method. Now we will. This does introduce a little bit of risk, but, this is hedged by the fact that we handle being unable to load the identity by just attempting to join, and if we think the identity might be invalid (e.g expired), we will also just attempt to join.